src/pytorch/pme/pmeCUDA.cu

```diff
@@ -0,0 +1,447 @@
+        int numAtoms = charges.size(0);
+        int numPairs = neighbors.size(1);
+        TensorOptions options = torch::TensorOptions().device(neighbors.device());
+        Tensor posDeriv = torch::zeros({numAtoms, 3}, options);
+        Tensor chargeDeriv = torch::zeros({numAtoms}, options);
+        int blockSize = 128;
+        int numBlocks = max(1, min(getMaxBlocks(), (numPairs+blockSize-1)/blockSize));
+        Tensor energy = torch::zeros(numBlocks*blockSize, options);
+        computeDirect<<<numBlocks, blockSize, 0, stream>>>(get_accessor<float, 2>(positions), get_accessor<float, 1>(charges),
+            get_accessor<int, 2>(neighbors), get_accessor<float, 2>(deltas), get_accessor<float, 1>(distances),
+            get_accessor<int, 2>(exclusions), get_accessor<float, 2>(posDeriv), get_accessor<float, 1>(chargeDeriv),
+            get_accessor<float, 1>(energy), alpha.toDouble(), coulomb.toDouble());
+
+        // Store data for later use.
+
+        ctx->save_for_backward({posDeriv, chargeDeriv});
+        return {torch::sum(energy)};
+    }
+
+    static tensor_list backward(AutogradContext *ctx, tensor_list grad_outputs) {
+        auto saved = ctx->get_saved_variables();
+        Tensor posDeriv = saved[0];
+        Tensor chargeDeriv = saved[1];
+        torch::Tensor ignore;
+        return {posDeriv*grad_outputs[0], chargeDeriv*grad_outputs[0], ignore, ignore, ignore, ignore, ignore, ignore};
+    }
+};
+
+class PmeFunctionCuda : public Function<PmeFunctionCuda> {
+public:
+    static Tensor forward(AutogradContext *ctx,
+                          const Tensor& positions,
+                          const Tensor& charges,
+                          const Tensor& box_vectors,
+                          const Scalar& gridx,
+                          const Scalar& gridy,
+                          const Scalar& gridz,
+                          const Scalar& order,
+                          const Scalar& alpha,
+                          const Scalar& coulomb,
+                          const Tensor& xmoduli,
+                          const Tensor& ymoduli,
+                          const Tensor& zmoduli) {
+        const auto stream = c10::cuda::getCurrentCUDAStream(positions.get_device());
+        const c10::cuda::CUDAStreamGuard guard(stream);
+        int numAtoms = positions.size(0);
+        int pmeOrder = (int) order.toInt();
+        int gridSize[3] = {(int) gridx.toInt(), (int) gridy.toInt(), (int) gridz.toInt()};
+        float sqrtCoulomb = sqrt(coulomb.toDouble());
+
+        // Spread the charge on the grid.
+
+        TensorOptions options = torch::TensorOptions().device(positions.device());
+        Tensor realGrid = torch::zeros({gridSize[0], gridSize[1], gridSize[2]}, options);
+        int blockSize = 128;
+        int numBlocks = max(1, min(getMaxBlocks(), (numAtoms+blockSize-1)/blockSize));
+        TORCH_CHECK(pmeOrder == 4 || pmeOrder == 5, "Only pmeOrder 4 or 5 is supported with CUDA");
```

Comment:
Why this restriction?

Reply:
The order is a template parameter to the kernel, so we have to specify all supported values at compile time.  4 and 5 are the only commonly used orders.

Reply:
Oh right, it is not a template CPU side.
If you think its worth, I usually dispatch things like that with something similar to this:
```c++
auto spread_kernel = spreadCharge<1>;
switch(pmeOrder){
       case 2: spread_kernel = spreadCharge<2>;
       case 3: spread_kernel = spreadCharge<3>;
       case 4: spread_kernel = spreadCharge<4>;
       ...
}
spread_kernel<<<...>>>(/*parameters*/)
```
Avoids having to copy all arguments.

Reply:
Good idea, except you're missing the `break` statements.  (I tend to avoid `switch` in C++ because it's so easy to make that mistake.)

Reply:
Done.